src/Microsoft.VisualStudio.Threading/JoinableTaskDependencyGraph.cs

```diff
@@ -219,6 +220,63 @@ internal static HashSet<JoinableTask> GetDependentTasksFromCandidates(IEnumerabl
             return results;
         }
 
+        /// <summary>
+        /// Computes dependency graph to clean up all potential unreachable dependents items.
+        /// </summary>
+        /// <param name="syncTask">A thread blocking sychornizing task.</param>
+        /// <param name="allReachableNodes">Returns all reachable nodes in the connected dependency graph, if unreachable dependency is found.</param>
+        /// <returns>True if it removes any unreachable items.</returns>
+        internal static bool CleanUpPotentialUnreachableDependentItems(JoinableTask syncTask, [MaybeNullWhen(false)] out HashSet<IJoinableTaskDependent>? allReachableNodes)
+        {
+            Requires.NotNull(syncTask, nameof(syncTask));
+
+            // a set of tasks may form a dependent loop, so it will make the reference count system
+            // not to work correctly when we try to remove the synchronous task.
+            // To get rid of those loops, if a task still tracks the synchronous task after reducing
+            // the reference count, we will calculate the entire reachable tree from the root.  That will
+            // tell us the exactly tasks which need track the synchronous task, and we will clean up the rest.
+            HashSet<IJoinableTaskDependent>? possibleUnreachableItems = syncTask.PotentialUnreachableDependents;
+            if (possibleUnreachableItems is object)
+            {
+                var reachableNodes = new HashSet<IJoinableTaskDependent>();
+                var syncTaskItem = (IJoinableTaskDependent)syncTask;
+
+                JoinableTaskDependentData.ComputeSelfAndDescendentOrJoinedJobsAndRemainTasks(syncTaskItem, reachableNodes, possibleUnreachableItems);
+
+                syncTask.PotentialUnreachableDependents = null;
+                allReachableNodes = reachableNodes;
+
+                // force to remove all invalid items
+                if (possibleUnreachableItems.Count > 0)
+                {
+                    JoinableTaskDependentData.RemoveUnreachableDependentItems(syncTask, possibleUnreachableItems, reachableNodes);
+
+                    return true;
+                }
+            }
+            else
+            {
+                allReachableNodes = null;
+            }
+
+            return false;
```

Comment:
`allReachableNodes` may yet be non-null here. Is that intentional?
Path: L239 is true, then L250 is false.

Reply:
originally thought it might be useful. But since we are not using it,  not returning it keeps things simple. I will clear it.